File: leetcode/medium/add-two-numbers.cpp
// https://leetcode.com/problems/add-two-numbers/
//
// Complexity: runtime O(m+n), space O(1) [output not considered].

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(const ListNode* l1, const ListNode* l2) const
    {
        ListNode* head = nullptr, end = nullptr;
        bool carry = false;

        while (l1 || l2 || carry) {
            const int digitSum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0)
                                 + (carry ? 1 : 0);
            carry = digitSum > 9;

            const auto newNode = new ListNode(digitSum % 10);

            if (!head)
                head = end = newNode;
            else
                end = end->next = newNode;

            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }

        return head;
    }
};

// Alternative Solution: shorter version (slightly harder to read).
//
// Many ways to make this even shorter (e.g: comma expressions), but this isn't
// code golf (or is it?)
//
// Complexity: runtime O(m+n), space O(1) [output not considered].
/*
class Solution {
public:
    ListNode* addTwoNumbers(const ListNode* l1, const ListNode* l2) const
    {
        ListNode* head = nullptr, * end = nullptr;
        int accum = 0;

        while (l1 || l2 || accum > 0) {
            accum += (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
            const auto newNode = new ListNode(accum % 10);

            if (!head)
                head = end = newNode;
            else
                end = end->next = newNode;

            accum /= 10;
            l1 = l1 ? l1->next : nullptr;
            l2 = l2 ? l2->next : nullptr;
        }

        return head;
    }
};
*/

File: leetcode/medium/add-two-numbers.rs
// https://leetcode.com/problems/add-two-numbers/
//
// Complexity: runtime O(m+n), space O(1).

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut result_fake_root = ListNode::new(0);
        let mut result_end = &mut result_fake_root;

        let (mut n1, mut n2) = (&l1, &l2);
        let mut carry = 0;

        while n1.is_some() || n2.is_some() || carry != 0 {
            let mut total = carry;
            if let Some(n) = n1 {
                total += n.val;
                n1 = &n.next;
            }
            if let Some(n) = n2 {
                total += n.val;
                n2 = &n.next;
            }

            result_end.next = Some(Box::new(ListNode::new(total % 10)));
            result_end = result_end.next.as_mut().unwrap();
            carry = total / 10;
        }

        result_fake_root.next
    }
}
